File:  LeetCode热题HOT100/1- 两数之和/main.c

#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
int *twoSum(int *nums, int numsSize, int target, int *returnSize)
{
	int max = -1;
	for (int i = 0; i < numsSize; i++)
	{
		if (max == -1)
		{
			max = nums[i];
		}
		else
		{
			max = (nums[i] > max) ? nums[i] : max;
		}
	}
	size_t alloc_sz = max / sizeof(uint8_t);
	if (alloc_sz % sizeof(uint8_t) != 0)
	{
		++alloc_sz;
	}
	int *result = calloc(*returnSize,sizeof(int));
	uint8_t *bits = calloc(alloc_sz, sizeof(uint8_t));
	for (int i = 0; i < numsSize; i++)
	{
		int index = nums[i] >> 3;
		bits[index] |= (1 << (nums[i] & 7));
	}
	bool flag = false;
	int left = -1;
	for (int i = 0; i < numsSize; i++)
	{
		if (!flag)
		{
			left = target - nums[i];
			int index = left >> 3;
			int ret = bits[index] & (1 << (left & 7));
			if (ret == 1)
			{
				result[0]=i;
				flag = true;
			}
		}else {
			if(nums[i]==left) {
				result[1]=i;
				break;
			}
		}
	}
	if (bits != NULL)
	{
		free(bits);
	}
	return result;
}

File:  LeetCode热题HOT100/1- 两数之和/main.go
package main

import (
	"fmt"
)

func twoSum(nums []int, target int) []int {
	res := make([]int, 0)
	dict := make(map[int][]int)
	for i, v := range nums {
		if dict[v] == nil {
			dict[v] = make([]int, 0)
		}
		dict[v] = append(dict[v], i)
		if dict[target-v] != nil {
			if len(dict[target-v]) > 1 {
				res = dict[target-v][:2]
			} else {
				if dict[v][0] != dict[target-v][0] {
					res = append(res, dict[v][0], dict[target-v][0])
					break
				}
			}
		}
	}
	/*
		for value, indexes := range dict {
			left := target - value
			if dict[left] != nil {
				if len(dict[left]) > 1 {
					res = indexes[:2]
				} else {
					if indexes[0] != dict[left][0] {
						res = append(res, indexes[0], dict[left][0])
						break
					}
				}

			}
		}

	*/
	return res
}
func main() {
	fmt.Println(twoSum([]int{2, 7, 11, 15}, 9))
	fmt.Println(twoSum([]int{3, 2, 4}, 6))
	fmt.Println(twoSum([]int{3, 3}, 6))
}
